subsys/dfu/dfu_target/src/dfu_target_smp.c

```diff
@@ -0,0 +1,342 @@
+{
+	ARG_UNUSED(cb);
+	if (img_num < 0) {
+		return -ENOENT;
+	}
+
+	if (dfu_target_reset_cb) {
+		upload_state.image_num = 2;
+	} else {
+		upload_state.image_num = (uint32_t)img_num;
+	}
+
+	upload_state.image_size = file_size;
+	upload_state.offset = 0;
+
+	return img_mgmt_client_upload_init(&img_gr_client, file_size, upload_state.image_num, NULL);
+}
+
+int dfu_target_smp_offset_get(size_t *out)
+{
+	*out = upload_state.offset;
+	return 0;
+}
+
+static int dfu_target_smp_enable(void)
+{
+	if (dfu_target_reset_cb && !recovery_mode_active) {
+		if (dfu_target_reset_cb()) {
+			LOG_ERR("SMP server reset fail");
+			return -ENOENT;
+		}
+
+		if (dfu_target_smp_recovery_enable()) {
+			LOG_ERR("SMP Server Recovery mode fail");
+			dfu_target_reset_cb();
+			return -ENOENT;
+		}
+		recovery_mode_active = true;
+	}
+	return 0;
+}
+
+int dfu_target_smp_write(const void *const buf, size_t len)
+{
+	int err = 0;
+	struct mcumgr_image_upload upload_res;
+
+	err = dfu_target_smp_enable();
+	if (err) {
+		return err;
+	}
+
+	err = img_mgmt_client_upload(&img_gr_client, buf, len, &upload_res);
+	if (err) {
+		return err;
+	}
+
+	return img_upload_res_update(&upload_res);
+}
+
+int dfu_target_smp_done(bool successful)
+{
+	int rc;
+	struct mcumgr_image_state res_buf;
+
+	rc = dfu_target_smp_enable();
+	if (rc) {
+		return rc;
+	}
+
+	if (successful) {
+		/* Read Image list */
+		LOG_INF("MCUBoot SMP image download ready.");
+		image_count = 0;
+		rc = img_mgmt_client_state_read(&img_gr_client, &res_buf);
+		if (rc) {
+			goto end;
+		}
+		rc = img_state_res_update(&res_buf);
+
+	} else {
+		/* Erase Image if recovery mode is not activated */
+		LOG_INF("MCUBoot SMP image upgrade aborted %d", upload_state.image_num);
+		if (!dfu_target_reset_cb) {
+			rc = img_mgmt_client_erase(&img_gr_client, upload_state.image_num);
+		}
+	}
+end:
+	upload_state.image_size = 0;
+	upload_state.offset = 0;
+
+	return rc;
+}
+
+static struct mcumgr_image_data *discover_secondary_image_info(void)
+{
+	if (!image_count) {
+		return NULL;
+	}
+
+	for (int i = 0; i < image_count; i++) {
+		if (image_info[i].slot_num == 0) {
+			continue;
+		}
+		return &image_info[i];
+	}
+
+	return NULL;
+}
+
+int dfu_target_smp_schedule_update(int img_num)
+{
+	int err = 0;
+	struct mcumgr_image_state res_buf;
+	struct mcumgr_image_data *secondary_image;
+
+	/* Discover allways Secondary image */
+	secondary_image = discover_secondary_image_info();
+	/*  Test functionality here */
+	if (!secondary_image) {
+		LOG_ERR("Secondary Image info not available");
+		return -ENOENT;
+	}
+
+	err = dfu_target_smp_enable();
+	if (err) {
+		return err;
+	}
+
+	LOG_INF("MCUBoot image-%d upgrade scheduled. "
+		"Reset device to apply",
+		secondary_image->slot_num);
+
+	err = img_mgmt_client_state_write(&img_gr_client, secondary_image->hash, false, &res_buf);
+	if (err) {
+		if (!dfu_target_reset_cb) {
+			img_mgmt_client_erase(&img_gr_client, upload_state.image_num);
+		}
```

Comment:
Not sure this really needs to be done or gated on if there is a reset

Reply:
OK. I remove that.

Reply:
What is the resolution on this? If an image transfer is incomplete, then it won't be found in e.g. image listings because the data is not valid, so I don't see the point in erasing it if you'd do this again when you start a new transfer anyway